specs/fault-proof.md

```diff
@@ -0,0 +1,412 @@
+
+Each of these 3 components may have different implementations, which can be combined into different proof stacks,
+and contribute to proof diversity when resolving a dispute.
+
+"Stateless execution" of the program, and its individual instructions, refers to reproducing
+the exact same computation by authenticating the inputs with a [Pre-image Oracle][oracle].
+
+![Diagram of Program and VM architecture](./assets/fault-proof.svg)
+
+## Pre-image Oracle
+
+[oracle]: #Pre-image-Oracle
+
+The pre-image oracle is the only form of communication between
+the [Program] (in the [Client](#client) role) and the [VM] (in the [Server](#server) role).
+
+The program uses the pre-image oracle to query any input data that is understood to be available to the user:
+
+- The initial inputs to bootstrap the program. See [Bootstrapping](#bootstrapping).
+- External data not already part of the program code. See [Pre-image types](#pre-image-types).
+
+The communication happens over a simple request-response wire protocol:
+
+- The client initiates a request for data by writing a `bytes32` [pre-image key](#pre-image-key-types).
+- The
+
+### Pre-image key types
+
+Pre-images are identified by a `bytes32` type-prefixed key:
+
+- The first byte identifies the type of request.
+- The remaining 31 bytes identify the pre-image key.
+
+#### Type `0`: Zero key
+
+The zero prefix is illegal. This ensures all pre-image keys are non-zero,
+enabling storage lookup optimizations and avoiding easy mistakes with invalid zeroed keys in the EVM.
+
+#### Type `1`: Local key
+
+Information specific to the dispute: the remainder of the key may be an index, a string, a hash, etc.
+Only the contract(s) managing this dispute instance may serve the value for this key:
+it is localized and contex-dependent.
+
+This type of key is used for program bootstrapping, to identify the initial input arguments by index or name.
+
+#### Type `2`: Global keccak256 key
+
+This type of key uses a global pre-image store contract, and is fully context-independent and permissionless.
+I.e. every key must have a single unique value, regardless of chain history or time.
+
+Using a global store reduces duplicate pre-image registration work,
+and avoids unnecessary contract redeployments per dispute.
+
+This global store contract should be non-upgradeable.
+
+Since `keccak256` is a safe 32-byte hash input, the first bit is overwritten with a `2` to derive the key,
+while keeping the rest of the key "readable" (matching the original hash).
+
+#### Type `3`: Global generic key
+
+Reserved. This scheme allows for unlimited application-layer pre-image types without fault-proof VM redeployments.
+
+This is a generic version of a global key store: `key = 0x03 ++ keccak256(x, sender)[1:]`, where:
+
+- `key` is a `bytes32`, which can be a hash of an arbitrary-length type of cryptographically secure commitment.
+- `sender` is a `bytes32` identifying the pre-image inserter address (left-padded to 32 bytes)
```

Comment:
What's `x` in this case? I think maybe it's actually being described as `key`.

Reply:
Mixed some things up here, will fix.